tools/crm_mon.c

```diff
@@ -896,6 +899,88 @@ set_fencing_options(int level)
     }
 }
 
+/* Before trying to connect to fencer or cib check for state of
+   pacemakerd - just no sense in trying till pacemakerd has
+   taken care of starting all the sub-processes
+
+   Only noteworthy thing to show here is when pacemakerd is
+   waiting for startup-trigger from SBD.
+ */
+static void
+pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
+                    enum pcmk_ipc_event event_type, crm_exit_t status,
+                    void *event_data, void *user_data)
+{
+    pcmk_pacemakerd_api_reply_t *reply = event_data;
+    enum pcmk_pacemakerd_state *state =
+        (enum pcmk_pacemakerd_state *) user_data;
+
+    /* we are just interested in the latest reply */
+    *state = pcmk_pacemakerd_state_invalid;
+
+    switch (event_type) {
+        case pcmk_ipc_event_reply:
+            break;
+
+        default:
+            return;
+    }
+
+    if (status != CRM_EX_OK) {
+        out->err(out, "Bad reply from pacemakerd: %s",
+                 crm_exit_str(status));
+        return;
+    }
+
+    if (reply->reply_type != pcmk_pacemakerd_reply_ping) {
+        out->err(out, "Unknown reply type %d from pacemakerd",
+                 reply->reply_type);
+    } else {
+        if ((reply->data.ping.last_good != (time_t) 0) &&
+            (reply->data.ping.status == pcmk_rc_ok)) {
+            *state = reply->data.ping.state;
+        }
+    }
+}
+
+static int
+pacemakerd_status(void)
+{
+    int rc = pcmk_rc_ok;
+    pcmk_ipc_api_t *pacemakerd_api = NULL;
+    volatile enum pcmk_pacemakerd_state state = pcmk_pacemakerd_state_invalid;
+
+    rc = pcmk_new_ipc_api(&pacemakerd_api, pcmk_ipc_pacemakerd);
+    if (pacemakerd_api == NULL) {
+        out->err(out, "Could not connect to pacemakerd: %s",
+                 pcmk_rc_str(rc));
+        /* this is unrecoverable so return with rc we have */
+        return rc;
+    }
+    pcmk_register_ipc_callback(pacemakerd_api, pacemakerd_event_cb, (void *) &state);
+    if ((rc = pcmk_connect_ipc(pacemakerd_api, pcmk_ipc_dispatch_poll)) == pcmk_rc_ok) {
+        if ((rc = pcmk_pacemakerd_api_ping(pacemakerd_api, crm_system_name)) == pcmk_rc_ok) {
+            if ((rc = pcmk_poll_ipc(pacemakerd_api, options.reconnect_msec/2)) == pcmk_rc_ok) {
+                pcmk_dispatch_ipc(pacemakerd_api);
+                rc = -ENOTCONN;
+                switch (state) {
+                    case pcmk_pacemakerd_state_running:
+                        rc = pcmk_ok;
```

Comment:
pcmk_rc_ok here and last line below

Reply:
Crap switched around between old and new style return-codes 2 times - and forgot those last time.